File: ai_diagnostic/bayesian.py
"""
Bayesian diagnostic engine.

Computes posterior probabilities P(Cause | evidence) using the
likelihood tables and evidence vectors defined in the paper.
"""

import math
from dataclasses import dataclass, field
from typing import Dict, List, Optional, Tuple

from .evidence import Evidence, EVIDENCE_PARAMS
from .likelihoods import (
    LikelihoodTable,
    FAILURE_CLASSES,
    FAILURE_CLASS_LABELS,
)


@dataclass
class DiagnosisResult:
    """Result of a Bayesian diagnostic computation."""

    posteriors: Dict[str, float]
    evidence_vector: Evidence
    diagnosis: str
    confidence: float
    needs_investigation: bool
    log_likelihoods: Dict[str, float]

    def __str__(self) -> str:
        lines = [
            "=" * 60,
            "  AI CONTROL FAILURE DIAGNOSIS",
            "=" * 60,
            "",
            "Evidence present:",
            self.evidence_vector.summary(),
            "",
            "-" * 60,
            "  Posterior Probabilities",
            "-" * 60,
        ]
        for cause in sorted(self.posteriors, key=self.posteriors.get, reverse=True):
            prob = self.posteriors[cause]
            label = FAILURE_CLASS_LABELS.get(cause, cause)
            bar = "#" * int(prob * 40)
            lines.append(f"  {label:<35s} {prob:>8.4f}  {bar}")

        lines.extend([
            "",
            "-" * 60,
            f"  DIAGNOSIS: {FAILURE_CLASS_LABELS.get(self.diagnosis, self.diagnosis)}",
            f"  CONFIDENCE: {self.confidence:.2%}",
        ])
        if self.needs_investigation:
            lines.append(
                "  ⚠ Confidence below threshold — further investigation recommended."
            )
        lines.append("=" * 60)
        return "\n".join(lines)

    def to_dict(self) -> dict:
        """Return results as a serializable dictionary."""
        return {
            "posteriors": self.posteriors,
            "diagnosis": self.diagnosis,
            "diagnosis_label": FAILURE_CLASS_LABELS.get(self.diagnosis, self.diagnosis),
            "confidence": self.confidence,
            "needs_investigation": self.needs_investigation,
            "log_likelihoods": self.log_likelihoods,
        }


class BayesianDiagnostic:
    """
    Bayesian diagnostic engine for AI control failures.

    Usage:
        diag = BayesianDiagnostic()
        result = diag.diagnose(evidence, prior_overrides=None)
        print(result)
    """

    def __init__(self, table: Optional[LikelihoodTable] = None):
        self.table = table or LikelihoodTable()

    def diagnose(
        self,
        evidence: Evidence,
        prior_overrides: Optional[Dict[str, float]] = None,
        confidence_threshold: float = 0.6,
    ) -> DiagnosisResult:
        """
        Compute posterior probabilities and return a diagnosis.

        Args:
            evidence: The observed evidence vector.
            prior_overrides: Optional dict overriding P(Cause) values.
            confidence_threshold: Minimum confidence to avoid flagging
                                  for further investigation.

        Returns:
            DiagnosisResult with posteriors, diagnosis, and confidence.
        """
        # Apply prior overrides
        table = self.table
        if prior_overrides:
            table = LikelihoodTable(
                priors={**self.table.priors, **prior_overrides},
                likelihoods=self.table.likelihoods,
            )

        # Compute log P(e | Cause) for numerical stability
        evidence_vec = evidence.to_vector()
        log_likelihoods: Dict[str, float] = {}
        log_priors: Dict[str, float] = {}

        for cause in FAILURE_CLASSES:
            log_priors[cause] = math.log(max(table.prior(cause), 1e-300))
            log_lik = 0.0
            for i, evid_name in enumerate(EVIDENCE_PARAMS):
                ei = evidence_vec[i]
                pe_c = table.p_evidence_given_cause(cause, evid_name)

                # For binary evidence (0 or 1):
                #   P(e=1 | C) = pe_c
                #   P(e=0 | C) = 1 - pe_c
                # For continuous evidence in (0,1):
                #   Use pe_c^ei * (1-pe_c)^(1-ei) as a interpolation
                if ei == 0.0:
                    log_lik += math.log(max(1.0 - pe_c, 1e-300))
                elif ei == 1.0:
                    log_lik += math.log(max(pe_c, 1e-300))
                else:
                    # Continuous: weighted geometric mean
                    log_lik += ei * math.log(max(pe_c, 1e-300)) + \
                               (1.0 - ei) * math.log(max(1.0 - pe_c, 1e-300))

            log_likelihoods[cause] = log_priors[cause] + log_lik

        # Normalize using log-sum-exp trick
        max_log = max(log_likelihoods.values())
        exp_shifted = {
            cause: math.exp(log_val - max_log)
            for cause, log_val in log_likelihoods.items()
        }
        total = sum(exp_shifted.values())
        posteriors = {
            cause: val / total for cause, val in exp_shifted.items()
        }

        # Determine diagnosis
        diagnosis = max(posteriors, key=posteriors.get)
        confidence = posteriors[diagnosis]
        needs_investigation = confidence < confidence_threshold

        return DiagnosisResult(
            posteriors=posteriors,
            evidence_vector=evidence,
            diagnosis=diagnosis,
            confidence=confidence,
            needs_investigation=needs_investigation,
            log_likelihoods=log_likelihoods,
        )

File: ai_diagnostic_rs/src/bayesian.rs
//! Bayesian diagnostic engine: computes P(Cause | evidence).

use crate::evidence::{Evidence, EVIDENCE_PARAMS};
use crate::likelihoods::{failure_class_label, LikelihoodTable, FAILURE_CLASSES};
use serde_json::{json, Map, Value};
use std::collections::HashMap;
use std::fmt;

const LOG_FLOOR: f64 = 1e-300;

#[derive(Debug, Clone)]
pub struct DiagnosisResult {
    /// Posterior per cause, ordered by FAILURE_CLASSES.
    pub posteriors: Vec<(String, f64)>,
    pub evidence_vector: Evidence,
    pub diagnosis: String,
    pub confidence: f64,
    pub needs_investigation: bool,
    /// Log P(e | Cause) + log prior per cause, ordered by FAILURE_CLASSES.
    pub log_likelihoods: Vec<(String, f64)>,
}

impl DiagnosisResult {
    pub fn posterior(&self, cause: &str) -> f64 {
        self.posteriors
            .iter()
            .find(|(c, _)| c == cause)
            .map(|(_, v)| *v)
            .unwrap_or(0.0)
    }

    fn sorted_posteriors(&self) -> Vec<(&str, f64)> {
        let mut pairs: Vec<(&str, f64)> = self
            .posteriors
            .iter()
            .map(|(c, v)| (c.as_str(), *v))
            .collect();
        pairs.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap_or(std::cmp::Ordering::Equal));
        pairs
    }
}

impl fmt::Display for DiagnosisResult {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut lines = vec![
            "=".repeat(60),
            "  AI CONTROL FAILURE DIAGNOSIS".to_string(),
            "=".repeat(60),
            String::new(),
            "Evidence present:".to_string(),
            self.evidence_vector.summary(),
            String::new(),
            "-".repeat(60),
            "  Posterior Probabilities".to_string(),
            "-".repeat(60),
        ];
        for (cause, prob) in self.sorted_posteriors() {
            let label = failure_class_label(cause);
            let bar = "#".repeat((prob * 40.0) as usize);
            lines.push(format!("  {label:<35} {prob:>8.4}  {bar}"));
        }
        lines.push(String::new());
        lines.push("-".repeat(60));
        lines.push(format!(
            "  DIAGNOSIS: {}",
            failure_class_label(&self.diagnosis)
        ));
        lines.push(format!("  CONFIDENCE: {:.2}%", self.confidence * 100.0));
        if self.needs_investigation {
            lines.push(
                "  ⚠ Confidence below threshold — further investigation recommended."
                    .to_string(),
            );
        }
        lines.push("=".repeat(60));
        write!(f, "{}", lines.join("\n"))
    }
}

impl DiagnosisResult {
    pub fn to_dict(&self) -> Map<String, Value> {
        let mut m = Map::new();
        m.insert(
            "posteriors".into(),
            Value::Object(
                self.posteriors
                    .iter()
                    .map(|(c, v)| (c.clone(), json!(v)))
                    .collect(),
            ),
        );
        m.insert("diagnosis".into(), json!(self.diagnosis));
        m.insert(
            "diagnosis_label".into(),
            json!(failure_class_label(&self.diagnosis)),
        );
        m.insert("confidence".into(), json!(self.confidence));
        m.insert("needs_investigation".into(), json!(self.needs_investigation));
        m.insert(
            "log_likelihoods".into(),
            Value::Object(
                self.log_likelihoods
                    .iter()
                    .map(|(c, v)| (c.clone(), json!(v)))
                    .collect(),
            ),
        );
        m
    }

    pub fn to_json_string(&self) -> String {
        Value::Object(self.to_dict()).to_string()
    }
}

#[derive(Debug, Clone, Default)]
pub struct BayesianDiagnostic {
    pub table: LikelihoodTable,
}

impl BayesianDiagnostic {
    pub fn new() -> Self {
        BayesianDiagnostic {
            table: LikelihoodTable::new(),
        }
    }

    pub fn with_table(table: LikelihoodTable) -> Self {
        BayesianDiagnostic { table }
    }

    /// Compute posteriors over the five failure classes.
    ///
    /// `prior_overrides` selectively replaces P(Cause) values.
    pub fn diagnose(
        &self,
        evidence: &Evidence,
        prior_overrides: Option<&HashMap<String, f64>>,
        confidence_threshold: f64,
    ) -> Result<DiagnosisResult, String> {
        evidence.validate()?;

        let log_likelihoods = self.compute_log_joint(evidence, prior_overrides);

        let max_log = log_likelihoods
            .iter()
            .map(|(_, v)| *v)
            .fold(f64::NEG_INFINITY, f64::max);
        let exp_shifted: Vec<(String, f64)> = log_likelihoods
            .iter()
            .map(|(c, v)| (c.clone(), (v - max_log).exp()))
            .collect();
        let total: f64 = exp_shifted.iter().map(|(_, v)| v).sum();
        let posteriors: Vec<(String, f64)> = exp_shifted
            .into_iter()
            .map(|(c, v)| (c, v / total))
            .collect();

        let (diagnosis, confidence) = posteriors
            .iter()
            .fold(("", f64::MIN), |acc, (c, v)| if *v > acc.1 { (c.as_str(), *v) } else { acc });
        let diagnosis = diagnosis.to_string();
        let needs_investigation = confidence < confidence_threshold;

        Ok(DiagnosisResult {
            posteriors,
            evidence_vector: evidence.clone(),
            diagnosis,
            confidence,
            needs_investigation,
            log_likelihoods,
        })
    }

    /// log P(e | C) + log P(C) per cause; overrides applied to a cloned table.
    fn compute_log_joint(
        &self,
        evidence: &Evidence,
        prior_overrides: Option<&HashMap<String, f64>>,
    ) -> Vec<(String, f64)> {
        let ev_vec = evidence.to_vector();

        FAILURE_CLASSES
            .iter()
            .map(|cause| {
                let base_prior = self.table.prior(cause).unwrap_or(LOG_FLOOR);
                let prior = prior_overrides
                    .and_then(|ov| ov.get(*cause).copied())
                    .unwrap_or(base_prior);
                let log_prior = prior.max(LOG_FLOOR).ln();

                let log_lik: f64 = EVIDENCE_PARAMS
                    .iter()
                    .zip(ev_vec.iter())
                    .map(|(name, ei)| {
                        let pe_c = self
                            .table
                            .p_evidence_given_cause(cause, name)
                            .unwrap_or(0.0);
                        if *ei == 0.0 {
                            (1.0 - pe_c).max(LOG_FLOOR).ln()
                        } else if *ei == 1.0 {
                            pe_c.max(LOG_FLOOR).ln()
                        } else {
                            ei * pe_c.max(LOG_FLOOR).ln()
                                + (1.0 - ei) * (1.0 - pe_c).max(LOG_FLOOR).ln()
                        }
                    })
                    .sum();

                ((*cause).to_string(), log_prior + log_lik)
            })
            .collect()
    }
}
